src/Parser.hs

```diff
@@ -0,0 +1,299 @@
+module Parser (
+    parseScheme,
+    Token(..),
+    MetaNode(..)
+) 
+where
+
+import Control.Monad.State.Lazy
+import Data.Data
+import Data.Complex
+
+
+
+data Token  = Lambda | If | Set | POpen | PClose | Identifier String | Quote | ShortQuote | Integral Int | Rational Int Int
+            | Real Double | String String | Complex (Complex Double)  | Bool Bool | Char Char
+            | Dot | QuasiQuote | ShortQuasiQuote | Unquote | ShortUnquote | UnquoteSplice | CommentDatum
+            | ShortUnquoteSplice | Label Integer | LabelRef Integer | Define deriving (Eq, Show) 
+
+data MetaNode = LambdaNode [MetaNode] MetaNode [MetaNode] | PairNode MetaNode MetaNode 
+            | RealAtom Double | IntegralAtom Int | RationalAtom Int Int | EmptyAtom
+            | StringAtom String | ComplexAtom (Complex Double) | BoolAtom Bool | CharAtom Char 
+            | IdentifierAtom String | ApplicationNode MetaNode [MetaNode] 
+            | IfNode MetaNode MetaNode MetaNode | SetNode MetaNode MetaNode | DefineNode MetaNode MetaNode 
+            deriving (Eq, Show)
+
+parseScheme :: [Token] -> MetaNode
+parseScheme st = case runState (parseExpression "Scheme Program")  st of
+    (mn, []) -> mn
+    (_, t:_) -> error $ "Unexpected token " ++ show t ++  " not allowed in top level of program"
+
+push :: Token -> State [Token] ()
+push t = do
+    ts <- get
+    put (t:ts)
+
+pull :: String -> State [Token] Token
+pull context = do
+    ts <- get
+    case ts of 
+        (t:ts) -> do
+            put ts
+            return t
+        _ -> error $ "Unexpected end of token stream in " ++ context
+
+pullEq :: String -> Token -> State [Token] ()
+pullEq context c = do
```

Comment:
pullEq is a variant of pull - maybe you can reuse pull inside of pullEq

Reply:
Good idea